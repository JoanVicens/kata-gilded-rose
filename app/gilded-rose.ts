import { Item } from "./Item";

export class GildedRose {
  private static readonly MAX_ITEM_QUALITY = 50;
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  private incrementItemQuality(item: Item): void {
    if (item.quality >= GildedRose.MAX_ITEM_QUALITY) {
      return;
    }

    if (item.name === "Aged Brie" && item.sellIn <= 0) {
      item.quality += 2;

      return;
    }

    item.quality += 1;

    if (item.name == "Backstage passes to a TAFKAL80ETC concert") {
      if (item.sellIn < 11) {
        item.quality += 1;
      }
      if (item.sellIn < 6) {
        item.quality += 1;
      }

      if (item.quality > GildedRose.MAX_ITEM_QUALITY) {
        item.quality = GildedRose.MAX_ITEM_QUALITY;
      }
    }
  }

  private decrementItemQuality(item: Item): void {
    if (
      item.name === "Sulfuras, Hand of Ragnaros" ||
      item.name === "Aged Brie"
    ) {
      return;
    }

    if (
      item.name === "Backstage passes to a TAFKAL80ETC concert" &&
      item.sellIn < 0
    ) {
      item.quality = 0;
      return;
    }

    if (item.quality > 0) {
      item.quality -= 1;
    }
  }

  private decreaseItemSellIn(item: Item): void {
    if (item.name != "Sulfuras, Hand of Ragnaros") {
      item.sellIn = item.sellIn - 1;
    }
  }

  private updateItemQuality(item: Item): void {
    if (
      item.name != "Aged Brie" &&
      item.name != "Backstage passes to a TAFKAL80ETC concert" &&
      item.name != "Sulfuras, Hand of Ragnaros"
    ) {
      this.decrementItemQuality(item);
    } else if (item.quality < GildedRose.MAX_ITEM_QUALITY) {
      this.incrementItemQuality(item);
    }

    this.decreaseItemSellIn(item);

    if (item.sellIn < 0) {
      this.decrementItemQuality(item);
    }
  }

  updateQuality() {
    this.items.forEach((item) => this.updateItemQuality(item));

    return this.items;
  }
}
