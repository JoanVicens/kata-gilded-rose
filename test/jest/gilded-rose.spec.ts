import { Item } from "@/Item";
import { GildedRose } from "@/gilded-rose";

describe("Gilded Rose", () => {
  it("should do a snapshot for 2 days", function () {
    let items = [
      new Item("+5 Dexterity Vest", 10, 20), //
      new Item("Aged Brie", 2, 0), //
      new Item("Elixir of the Mongoose", 5, 7), //
      new Item("Sulfuras, Hand of Ragnaros", 0, 80), //
      new Item("Sulfuras, Hand of Ragnaros", -1, 80),
      new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
      new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
      new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
      // this conjured item does not work properly yet
      new Item("Conjured Mana Cake", 3, 6),
    ];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });

  it("should do a snapshot for 10 days", function () {
    let items = [
      new Item("+5 Dexterity Vest", 10, 20), //
      new Item("Aged Brie", 2, 0), //
      new Item("Elixir of the Mongoose", 5, 7), //
      new Item("Sulfuras, Hand of Ragnaros", 0, 80), //
      new Item("Sulfuras, Hand of Ragnaros", -1, 80),
      new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
      new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
      new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
      // this conjured item does not work properly yet
      new Item("Conjured Mana Cake", 3, 6),
    ];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });
});
