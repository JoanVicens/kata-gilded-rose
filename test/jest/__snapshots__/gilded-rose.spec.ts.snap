// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose should do a snapshot for 2 days 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  Item {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 4,
    "sellIn": 1,
  },
]
`;

exports[`Gilded Rose should do a snapshot for 10 days 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 10,
    "sellIn": 0,
  },
  Item {
    "name": "Aged Brie",
    "quality": 18,
    "sellIn": -8,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -5,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 35,
    "sellIn": 5,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 0,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -5,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 0,
    "sellIn": -7,
  },
]
`;
